library/core/src/main/java/com/google/android/exoplayer2/text/cea/Cea608Decoder.java

```diff
@@ -257,6 +257,16 @@
   // service bytes and drops the rest.
   private boolean isInCaptionService;
 
+  // Static counter to keep track of last CC rendered. This is used to force erase the caption when
+  // the stream does not explicitly send control codes to remove caption as specified by
+  // CEA-608 Annex C.9
+  private long ccTimeOutCounter = C.TIME_UNSET;
+  private boolean captionEraseCommandSeen = false;
+  // CEA-608 Annex C.9 propose that if no data are received for the selected caption channel within
+  // a given time, the decoder should automatically erase the caption. The time limit should be no
+  // less than 16 seconds
+  public static final int VALID_DATA_CHANNEL_TIMEOUT_MS = 16000;
```

Comment:
What happens when there is, for example, a song playing with captions "♫♫". Those captions are usually shown for a many seconds. Are those repeatedly sent, processed and kept on the screen correctly?

When restarting the timer, have you considered special cases, like:
- receiving a duplicate of a previous command
- receiving byte to a CEA608 mode that is not part of the captions, like TEXT or RDS mode
- receiving an invalid command
- when process_cc_data_flag is set to false
- receiving a value with incorrect CRC
- receiving bytes that are marked not to process.
- receiving bytes to a different channel, not the currently selected ones (like receiving CC3 bytes, but nothing for CC1)
- in ROLL_UP mode, when there is no one talking for a prolonged period of time, so no roll event happens.

Did you have a chance to run certification test cases with the change? Like the Sarnoff test cases for CEA608 and CEA708?

Reply:
Thank you @zsmatyas.

I don’t see the captions "♫♫”. Its sent once followed by null bytes (0x0, 0x0) and EOC. In this fix, the max displayed period is 16 seconds. If EOC arrives before the timeout, the caption is erased and timer is stopped. Otherwise the caption is erased after 16 second timeout.

Timer is started in Cea608Decoder.decode() function when the flag captionDataProcessed is set and caption mode is ROLL_UP and PAINT_ON

When restarting the timer, 
* receiving a duplicate of a previous command
> Cea608Decoder ignores duplicate commands in isRepeatedCommand(). In such case the timer continues to be in its current state. I.e. if already started it continues, otherwise timer is not started on duplicate command

* receiving byte to a CEA608 mode that is not part of the captions, like TEXT or RDS mode
> Timer is started only in ROLL_UP and PAINT_ON mode.
* receiving an invalid command
>  I believe as per CEA-608, there can’t be an invalid control word. Annex D.2 mentions - “Encoder manufacturers shall not insert these invalid control codes on line 21, field 1, as services designed for line 21, field 2, use these codes as valid data”. I don’t see any check done for invalid command in CeaDecoder path. So, for any reason if Cea608Decoer gets invalid command code the timer is started as it is done on normal Control codes. By the way, timer is not started on control words such as End_Of_Caption(EOC 14 2F), Erase_Displayed_Memory(EDM 14 2C), Eras_NonDisplay_Memory(ENM 14 2E) since these commands erase the captions.
* when process_cc_data_flag is set to false
    This is filtered at UserDataReader. If process_cc_data_flag is not set, the consumeCcData() function discards the ccDataBuffer. So the Cea608Decoder won’t see those triplets and the timer is not started
* receiving a value with incorrect CRC
> Please correct me in case I am missing something here. The cc_data syntax doesn’t describe CRC. I don’t seem to find CRC in CEA-608 either. 
* receiving bytes that are marked not to process.
> Could you please elaborate on this. Are you referring to cc_valid flag? If cc_valid flag is not set the, the timer is not started on reception of such triplets.  
* receiving bytes to a different channel, not the currently selected ones (like receiving CC3 bytes, but nothing for CC1)
> Timer is not started in such case.
* in ROLL_UP mode, when there is no one talking for a prolonged period of time, so no roll event happens.
> The previously displayed caption is erased after 16 seconds timeout if no roll event or no valid data is received (Captions are erased after 16 seconds even if null bytes (0x0, 0x0) are seen for 16 seconds). 

No, I did not run the certification tests with this change. I will have to check the availability of Sarnoff streams. 

Thanks,
 Sadashiva N.

Reply:
Thanks for your comments @zsmatyas - as you've contributed a fair amount of the CEA support to ExoPlayer it's helpful to hear your opinions.

From my reading of the spec this is an optional feature.

It seems there is a non-zero risk that this inadvertently removes a subtitle that was intended to be shown for longer than 16s. The benefit of this change is that invalid streams that are missing the End_Of_Caption control word don't have their subtitles stuck on screen for longer than intended.

Currently I'm inclined to keep the current behaviour and reject this PR, since it seems better for invalid data to result in "stuck" rendering than valid data to result in "early-removal".

> No, I did not run the certification tests with this change. I will have to check the availability of Sarnoff streams.

If you're able to do that it would be excellent. Unfortunately we don't have access to these test streams.

Reply:
Thanks Ian (@icbaker) Sada (@sneelavara) works with me here at TiVo, we of course have the Sarnaoff transport stream files for testing our STB's, we just need to work out how to get access to use these for this testing.  

BTW,  Since TiVo delivers broadcast content we have been mandated by the FCC to provide CC support since day one.  This clear logic has been in our STB's for years.  Very grateful for @zsmatyas insight on this, we will work to make this bullet proof before we ask you @icbaker to make it merge ready.

Reply:
I had the privilege of debugging many complaints for broken captions. My impression was that content providers (various live channels creating content or only subtitles to existing content) around North America seem to interpret the CEA608 captions in different ways - as originally it was targeting over-the-air analogue content and was not prepared for the use cases we have today.
I have met many issues that are caused by undefined values, unexpected ordering, ambiguous wording in the standard and similar - and the issue is not our interpretation of the expectation, but the interpretation of the content providers or the tools they use.

I agree that the commercials not having any subtitle commands interrupting normal content with correct captions cause issues and the "stuck" lines are annoying, but we could not propose a way that has more benefit than side effects and will correctly work with the thousands of untested channels we do not have access to at the moment. For example, if we test it and fix all issues with the current user base, some future application released in a county not yet tested but also using CEA608 still might have unexpected issues.

The Sarnoff test cases are very useful because they exercise the functionality in many unexpected ways. One of them was, for example, "flashing" caused by switching the visible and invisible buffers without any other commands received. This proposed code change might regress such test cases as it clears the caption buffers when the test case expects them not to be modified (only the visibility is flipped). 

One more use case to consider, when the normal content is interrupted by ads without captions: when returning to the normal content, the buffers for pending (but not visible) captions should have the content from before the advertisements were started, otherwise the first set of captions will not be present and the state of the caption decoder might be altered. But this is a general issue with today's solution of injecting commercials into content previously correctly working with 608 captions - you need to consider the timing and the state of the decoder. If the content provider does that correctly, we might expect them to start with a "clear" command.

The proposed change might probably be useful for many developers. But in general, I would suggest not merging this change without a flag to enable / disable the proposed "clear" functionality as most probably it should not be applied universally. One path forward could be have this feature turned off by default, so developer need to opt-in for the effects.

Reply:
@zsmatyas I'm sure.  Captions are often problematic.  The use case we see this most on is live authored streams (news channels like CNN etc).  We certainly welcome your help for a best solution to this, looks like you're living this problem too.